### ISSUE TITLE
Improve foldl performance on Interators.Filter

### ISSUE THREAD

**Issue by MasonProtter** [JuliaFolds/Transducers.jl#306]
So I noticed in this random benchmark that doing a `foldl` over a `Iterators.filter` is doing something inefficient:
```julia
using Transducers

min1(f, iter) = minimum(f, iter) 

min2(f, iter) = foldl(Map(f), iter) do min_so_far, x
    if x < min_so_far
        x
    else
        min_so_far
    end
end

function min3(f, iter) 
    min_so_far = f(first(iter))
    for x ∈ iter
        fx = f(x)
        if fx < min_so_far
            min_so_far = fx
        end 
    end
    min_so_far
end
```
```julia
let f = x -> x^2 + 2x + 100, iter = Iterators.filter(x -> x < 0, randn(1000))
    @btime min1($f, $iter) 
    @btime min2($f, $iter) 
    @btime min3($f, $iter)
end;

#+RESULTS:
:   2.202 μs (0 allocations: 0 bytes)
:   17.180 μs (1481 allocations: 23.14 KiB)
:   1.355 μs (0 allocations: 0 bytes)
```
it seems though that the problem is simply that Transducers.jl is not splitting the `Iterators.Filter` it got into a trandsducer and an iterator, and if I define a method that does so, things improve:
```julia
Base.foldl(f, xf::Transducer, fiter::Iterators.Filter) = foldl(f, Filter(fiter.flt) |> xf, fiter.itr)

let f = x -> x^2 + 2x + 100, iter = Iterators.filter(x -> x < 0, randn(1000))
    @btime min1($f, $iter) 
    @btime min2($f, $iter) 
    @btime min3($f, $iter)
end;

#+RESULTS:
:   2.213 μs (0 allocations: 0 bytes)
:   1.864 μs (0 allocations: 0 bytes)
:   1.228 μs (0 allocations: 0 bytes)
```

I had assumed this was something that Transducers.jl was already doing under the hood, but maybe I was mistaken. Would this method and similar ones for other iterators be welcome as a PR? If so, what file would that go in? `src/processes.jl`

**Comment by tkf** [comment]
Thanks for bringing this up! Actually, there is `eduction` that does this. But I haven't hooked it up yet in this code path. #310 should fix it.

**Comment by MasonProtter** [comment]
Right, I forgot to mention here that reducing (or worse, iterating) over an `eduction` appears to be even slower:
```julia
julia> let f = x -> x^2 + 2x + 100, iter = eduction(Filter(<(0)), randn(1000))  
           @btime min1($f, $iter) 
           @btime min2($f, $iter)
           @btime min3($f, $iter)
       end
  49.840 μs (2029 allocations: 71.48 KiB)
  31.609 μs (2032 allocations: 47.88 KiB)
  49.739 μs (2036 allocations: 71.88 KiB)
99.0000378215463
```
Will #310 fix that as well, or are they unrelated?

**Comment by tkf** [comment]
Yeah, prior to #310, you need `eduction(Filter(<(0)) |> Map(f), randn(1000))`.  This should be fixed by #310, too.

**Comment by tkf** [comment]
`minimum(eduction(_))` is fixed by #72
